crates/bindings-csharp/Runtime/Internal/ITable.cs

```diff
@@ -22,31 +22,51 @@ public class Enumerator(FFI.RowIter handle) : IDisposable
 
             public bool MoveNext()
             {
+                if (handle == FFI.RowIter.INVALID)
+                {
+                    return false;
+                }
+
                 uint buffer_len;
                 while (true)
                 {
                     buffer_len = (uint)buffer.Length;
-                    try
+                    var ret = (FFI.Errno)
+                        FFI._row_iter_bsatn_advance(handle, buffer, ref buffer_len);
+                    if (ret == FFI.Errno.EXHAUSTED)
                     {
-                        FFI._iter_advance(handle, buffer, ref buffer_len);
+                        handle = FFI.RowIter.INVALID;
                     }
-                    catch (BufferTooSmallException)
+                    switch (ret)
                     {
-                        buffer = new byte[buffer_len];
-                        continue;
+                        // Iterator exhausted, we're done.
+                        case FFI.Errno.EXHAUSTED:
+                        // Iterator advanced but not exhausted,
+                        // we'll need to advance the iterator in the next call to `MoveNext`.
+                        case FFI.Errno.OK:
+                            // Success! Copy over the row data to `Current` from the scratch `buffer`.
+                            Current = new byte[buffer_len];
+                            Array.Copy(buffer, 0, Current, 0, buffer_len);
+                            return buffer_len != 0;
+                        // Couldn't find the iterator, error!
+                        case FFI.Errno.NO_SUCH_ITER:
+                            throw new NoSuchIterException();
+                        // The scratch `buffer` is too small to fit a row / chunk.
+                        // Grow `buffer` and try again.
+                        case FFI.Errno.BUFFER_TOO_SMALL:
+                            buffer = new byte[buffer_len];
```

Comment:
Unlike the comment says, it doesn't grow the `buffer` here?

Reply:
Ahh I see, `buffer_len` has a new length at this point. Might be worth adding to the comment, it's easy to miss.